File: JS/0240_Search_A_2D_Matrix_II.js
var search = function(nums, target){
  let lhs = 0, rhs = nums.length
  while (lhs < rhs){
    let mid = Math.floor((lhs + rhs) / 2)
    if (nums[mid] >= target){
      rhs = mid
    }else{
      lhs = mid + 1
    }
  }
  return lhs
}

var searchMatrix = function(matrix, target) {
  for (let row of matrix){
    idx = search(row, target)
    if (idx < row.length && row[idx] == target){
      return true
    }
  }
  return false
};

File: Go/0240_Search_A_2D_Matrix_II.go
func search(nums []int, target int) int {
	lhs, rhs := 0, len(nums)
	for lhs < rhs {
		mid := (lhs + rhs) / 2
		if nums[mid] >= target {
			rhs = mid
		} else {
			lhs = mid + 1
		}
	}
	return lhs
}

func searchMatrix(matrix [][]int, target int) bool {
    for _, row := range matrix {
		idx := search(row, target)
		if idx < len(row) && row[idx] == target {
			return true
		}
	}
	return false
}
